### PR TITLE
AAInstr: Add parallel auto tuner

### PR DESCRIPTION
**Pull request by webmiche**

### CODE REVIEW

File: AAInstrumentation/optimization.py [hunk 1]
```diff
@@ -180,6 +180,71 @@ def autotuner_factory():
     return lambda driver: AutoTuningOptimizer(driver)
 
 
+class ParallelAutotuner(Optimizer):
+    """
+    This class implements the local Autotuner in a parallel way. It aims to
+    optimize the size of the final binary by sequentially trying out the queries
+    in the compilation. If a query reduces the size, it is fied to imprecise for
```
Comment (thetheodor) [inline AAInstrumentation/optimization.py]: its is fied?

File: AAInstrumentation/optimization.py [hunk 2]
```diff
@@ -180,6 +180,71 @@ def autotuner_factory():
     return lambda driver: AutoTuningOptimizer(driver)
 
 
+class ParallelAutotuner(Optimizer):
+    """
+    This class implements the local Autotuner in a parallel way. It aims to
```
Comment (thetheodor) [inline AAInstrumentation/optimization.py:191]: would it make sense to rename the class to `ParallelLocalAutotuner`?

File: AAInstrumentation/optimization.py [hunk 3]
```diff
@@ -180,6 +180,71 @@ def autotuner_factory():
...
+        self.batch_size = driver.proc_count
+
+    description = "Parallel AutoTuner"
+    batch_size: int = 0
+
+    def handle_batch_from_index(
+        self, file_name: Path, prefix: list[int], index: int
+    ) -> list[int]:
+        sizes = []
+        with Pool(self.batch_size) as pool:
```
Comment (thetheodor) [inline AAInstrumentation/optimization.py:215]: Is this a thread pool or a process pool?

I'm not sure if it matters a lot given that you are only dealing with subprocesses, but maybe it makes a difference

File: AAInstrumentation/optimization.py [hunk 4]
```diff
@@ -180,6 +180,71 @@ def autotuner_factory():
...
+        return sizes
+
+    def update_iteration(
+        self, file_name: Path, prefix: list[int], index: int, current_minimum: int
+    ) -> tuple[list[int], int, int]:
+        sizes = self.handle_batch_from_index(file_name, prefix, index)
+
+        for i, size in enumerate(sizes):
+            if size < current_minimum:
+                return prefix + [index + i], index + i, size
```
Comment (thetheodor) [inline AAInstrumentation/optimization.py:233]: I'm not sure if it makes sense, but I think the code would be much more readable if there were explicit types for these AA sequences, then also the return types won't be as cryptic as `tuple[list[int], int, int]`

**Review by thetheodor — APPROVED** [review]

File: AAInstrumentation/optimization.py (hunk 3 above)
Comment (webmiche) [reply AAInstrumentation/optimization.py:215]: I think it's a process pool, AFAIU. Do you think a thread tool would make more sense?

File: AAInstrumentation/optimization.py (hunk 4 above)
Comment (webmiche) [reply AAInstrumentation/optimization.py:233]: I decided to only this for this specific optimizer atm. I will file a follow up PR applying this to the rest, once this is merged.

**Pull request merged**